### ISSUE TITLE
"unreachable" nexthop

### ISSUE THREAD

**Issue by ozdanborne** [projectcalico/calico-bird#31]
Using master builds, k8s 1.6 on GCE

Pods don't get connectivity (to the k8s apiserver at least). 

Routing table shows an `unreachable` next hop
```
root@turk-kubeadm-xxx:/home/turk# ip route
default via 10.240.0.1 dev ens4 
10.240.0.1 dev ens4  scope link 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
unreachable 192.168.11.128/26  proto bird 
blackhole 192.168.133.0/26  proto bird 
```

Though `calicoctl node status` shows BGP is established.

Bird logs repeatedly show this message: `bird: KRT: Received route 0.0.0.0/0 with strange next-hop 10.240.0.1`

[node-logs.txt](https://github.com/projectcalico/bird/files/860123/node-logs.txt)

cc: @robbrockbank

**Comment by nelljerram** [comment]
Looking into this... I don't see anything obviously missing from the newly merged code, so my feeling is that this is going to take some time to understand and fix.  Therefore, if we need a quick fix for operation with k8s 1.6 on GCE, we should bear in mind the option of just reverting back to before the 1.6.3 merge.

**Comment by nelljerram** [comment]
On that point, do we know for sure if our BIRD v0.2.1 worked in the same scenario?

(I'm guessing we may have only just started testing k8s 1.6, so just possibly this might be a k8s 1.5->1.6 issue, instead of being a BIRD code-merge issue.)

**Comment by ozdanborne** [comment]
Dang. I encountered this issue on a system where I was repeatedly reinstalling kubernetes clusters. It was likely a result of stale etcd data, as etcd volume mounts and (if I am not careful) persists into the next kubernetes cluster. We should confirm that on a fresh cluster, master builds are working OK

**Comment by nelljerram** [comment]
We definitely do have a problem with the new code on GCE.

With a k8s deployment with one (non-system) pod created, routes on master with old BIRD code
are:

neil@neil-kubetest2-master:~$ ip r
default via 10.240.0.1 dev ens5
10.240.0.1 dev ens5  scope link
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown
blackhole 192.168.41.0/26  proto bird
192.168.71.0/26 via 10.240.0.28 dev tunl0  proto bird onlink
192.168.73.64/26 via 10.240.0.37 dev tunl0  proto bird onlink

Same but with merged BIRD code:

neil@neil-kubetest-master:~$ ip r
default via 10.240.0.1 dev ens5
10.240.0.1 dev ens5  scope link
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown
blackhole 192.168.11.128/26  proto bird
unreachable 192.168.210.192/26  proto bird
unreachable 192.168.230.0/26  proto bird

Also the merged BIRD code generates these logs:

bird: KRT: Received route 0.0.0.0/0 with strange next-hop 10.240.0.1

whereas the old BIRD code does not.

Pretty sure the problem is that we lost two changes like this in the merge:

```
-	  neighbor *ng = neigh_find2(&p->p, &rv->gw, rv->iface,
-				     (nh->rtnh_flags & RTNH_F_ONLINK) ? NEF_ONLINK : 0);
+	  neighbor *ng = neigh_find2(&p->p, &rv->gw, rv->iface, NEF_ONLINK);
```

(introduced originally by https://github.com/projectcalico/bird/commit/6af4e30d3fc)

I could just fix those, but think I need to review the merge more generally for conflict resolutions.

**Comment by nelljerram** [comment]
Oh, one more observation to report:

I can fix the 'unreachable' routes on the rig with the merged BIRD code just by reprogramming the node's default route like this:

    neil@neil-kubetest-master:~$ sudo ip r del default via 10.240.0.1 dev ens5; sudo ip r add default via 10.240.0.1 dev ens5 onlink

This confirms the association with the code changes mentioned above.

**Comment by nelljerram** [comment]
This issue was fixed initially by reverting the merge of #30, and then later by #33.
